### DISCUSSION TITLE
Processing of external audio

### DISCUSSION THREAD

**Discussion by nilspupils** [mcguirepr89/BirdNET-Pi#118] [Q&A] [answered]
Hi guys, 
when processing external audio files (e.g. from dropped off audio devices), what is the best way to do it? I tried to start analyze.py from the command line but could not get it to work: 

`birdnetpi:~/BirdNET-Pi$ python3 analyze.py --i 'example/XC558716 - Soundscape.mp3' --lat 35.4244 --lon -120.7463 --week 18
Traceback (most recent call last):
  File "/home/pi/BirdNET-Pi/analyze.py", line 8, in <module>
    import tflite_runtime.interpreter as tflite
ModuleNotFoundError: No module named 'tflite_runtime'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pi/BirdNET-Pi/analyze.py", line 10, in <module>
    from tensorflow import lite as tflite
ModuleNotFoundError: No module named 'tensorflow'`

Thanks and sorry if this is a really stupid question. ;)
Best, Nils

**Comment by mcguirepr89** [comment, marked as answer]
Howdy, @nilspupils!

`analyze.py` requires the `birdnet` python virtual environment in order to locate all of its site-packages. 
To activate the python virtual environment, issue:
```
source ~/BirdNET-Pi/birdnet/bin/activate
```
You'll notice that command changes your bash prompt to indicate the virtual environment.

From there, you can run the `analyze.py` script and it should no longer fuss.

Let me know if it does still fuss, and I'll take a closer look.

My best!

**Comment by CaiusX** [comment]
Hi @nils

Use the python virtual environment under birdnet for the tflite and other libraries to be accessible

**Reply by nilspupils** [reply]
Thanks to both of you. I will check this out!

**Reply by nilspupils** [reply]
This worked surprisingly well, but only after i replaced the analyze.py with the unedited original.
Many thanks!

**Reply by mcguirepr89** [reply]
Yeah that particular file underwent several changes today, so timing could have been an issue!

**Comment by mcguirepr89** [comment]
@nilspupils Hello, again!

I've been thinking on this a bit more and remembered that I put together a "drop in" feature a while back. It's kind of a hidden feature, but the idea is that the `birdnet_analysis.sh` script checks for `WAV` files in the `$RECS_DIR`, so you could just put `WAV` files in the `/home/pi/BirdSongs` directory and they would automatically get processed just like the recordings that BirdNET-Pi makes.

Thinking on this a bit more, however, I think I'll simply add a new directory, `$RECS_DIR/DropIns` where it is more clear that you can just upload `WAV` files to that directory and they will get processed within the `birdnet_analysis.service`.

I would love it if you would be my tester on that feature so that I can try to make it work consistently. This would allow you to just upload files using the File Manager.

Let me know what you think?
My best!
Patrick

**Reply by Molokai** [reply]
This feature, along with the ability to live stream audio, seems to indicate to me that one COULD run just  the recording + pulseaudio on low-powered devices (like maybe even a rPi Zero W - and stream the audio to a full BirdNet.Pi install for processing (or perhaps some more powerful computer or docker container)

**Reply by mcguirepr89** [reply]
Yes -- the first version of this software utilized that arrangement. The design (that I still have code for) uses SSHFS to mount the recording device's input to the processing device's file structure. That way, a $5 RPi0W can be outside sending data back to an RPi4B that is safely inside. This setup does require a relatively small network to keep things connected. If you are interested in that code, I could probably dig it up, but is definitely available in some older iterations of [BirdNET-system](https://github.com/mcguirepr89/BirdNET-system)

**Reply by nilspupils** [reply]
@mcguirepr89: Sounds like a great idea! Let's do it! Since my audio-drop-off device will produce mp3, would it be possible to include these also? The test yesterday was with a mp3-file and it worked just fine! There will have to be an option for date/time to go with the audio file so that the database entry will be right.
Best, 
Nils

**Reply by mcguirepr89** [reply]
I will need to adjust things just a bit to look for `mp3`s, but that won't be any trouble at all. Handling the timing and date stuff will be the trickiest part. I can say that `analyze.py` does check the filename for the date/time info, so if you can get that into the drop-off recording file names, that would be amazing. If not, we can probably come up with some other means to get things to work.

**Reply by nilspupils** [reply]
Probably the best solution to have it in the filename! Easy enough to do and no mess with other files and notes...
